python/lsst/pipe/base/executionButlerBuilder.py

```diff
@@ -39,6 +40,52 @@
 DataSetTypeMap = Mapping[DatasetType, Set[DataCoordinate]]
 
 
+def _validate_dataset_type(
+    candidate: DatasetType, previous: dict[Union[str, DatasetType], DatasetType]
+) -> DatasetType:
+    """Check the dataset types and return a consistent variant if there are
+    different compatible options.
+
+    Parameters
+    ----------
+    candidate : `lsst.daf.butler.DatasetType`
+        The candidate dataset type.
+    previous : `dict` [Union[`str`, `DatasetType`], `DatasetType`]
+        Previous dataset types found, indexed by name and also by
+        dataset type. The latter provides a quick way of returning a
+        previously checked dataset type.
+
+    Returns
+    -------
+    datasetType : `lsst.daf.butler.DatasetType`
```

Comment:
Please clarify that it returns `candidate` (which is a DatasetType), or raises if `candidate` is not compatible

Reply:
It doesn't always return the supplied candidate. It will mostly return it but the purpose of this function is to sometimes have to return something different. The problem being solved is that if your pipeline declares a metadata dataset as input with PropertySet storage class and the graph has been built with the metadata dataset (that is automatically being injected as an output of every task) being a TaskMetadata one, you sometimes are given a TaskMetadata dataset type and sometimes a PropertySet. This function guarantees to return a consistent dataset type.